dags/veda_data_pipeline/utils/transfer.py

```diff
@@ -0,0 +1,100 @@
+import os
+import re
+
+import boto3
+
+
+def assume_role(role_arn, session_name="veda-data-airflow_s3-discovery"):
+    sts = boto3.client("sts")
+    print(f"Assuming role: {role_arn}")
+    credentials = sts.assume_role(
+        RoleArn=role_arn,
+        RoleSessionName=session_name,
+    )
+    creds = credentials["Credentials"]
+    return {
+        "aws_access_key_id": creds["AccessKeyId"],
+        "aws_secret_access_key": creds.get("SecretAccessKey"),
+        "aws_session_token": creds.get("SessionToken"),
+    }
+
+
+def get_s3_resp_iterator(bucket_name, prefix, s3_client, page_size=1000):
+    """
+    Returns an s3 paginator.
+    :param bucket_name: The bucket.
+    :param prefix: The path for the s3 granules.
+    :param s3_client: Initialized boto3 S3 client
+    :param page_size: Number of records returned
+    """
+    s3_paginator = s3_client.get_paginator("list_objects")
+    return s3_paginator.paginate(
+        Bucket=bucket_name, Prefix=prefix, PaginationConfig={"page_size": page_size}
+    )
+
+
+def get_matching_files(s3_client, bucket, prefix, regex_pattern):
+    matching_files = []
+
+    response = s3_client.list_objects_v2(Bucket=bucket, Prefix=prefix)
+    while True:
+        for obj in response["Contents"]:
+            file_key = obj["Key"]
+            if re.match(regex_pattern, file_key):
+                matching_files.append(file_key)
+
+        if "NextContinuationToken" in response:
+            response = s3_client.list_objects_v2(
+                Bucket=bucket,
+                Prefix=prefix,
+                ContinuationToken=response["NextContinuationToken"],
+            )
+        else:
+            break
+
+    return matching_files
+
+
+def transfer_files_within_s3(
+    s3_client, origin_bucket, matching_files, destination_bucket, collection
+):
+    for file_key in matching_files:
+        filename = file_key.split("/")[-1]
+        target_key = f"{collection}/{filename}"
+        copy_source = {"Bucket": origin_bucket, "Key": file_key}
+        s3_client.copy_object(
+            CopySource=copy_source, Bucket=destination_bucket, Key=target_key
+        )
```

Comment:
Can we check if the file already exists in the target bucket (same filename + e-tag) before copying?